### ISSUE TITLE
3.x Allow removal of single event listeners

### ISSUE THREAD

**Issue by davidyell** [cakephp/cakephp#6015]
Currently I don't believe that it's possible to remove a single event listener without having the same instance available.

This means that if you are emitting events and someone else wants to replace your listener they must remove all the attached events and readd the ones they want.

So the way I currently understand the events api is thus,
```php
// ExampleTable.php
public function initialize(array $config)
{
    //etc
    $this->eventManager()->off('Some.CustomEvent');

    $listener = new App\Event\DefaultListener();
    $this->eventManager()->on($listener);

    $listener = new App\Event\ReplacementListener();
    $this->eventManager()->on($listener);
```

It would be nice to have an api similar to this I think,

```php
// ExampleTable.php
public function initialize(array $config)
{
    //etc
    $this->eventManager()->off('Some.CustomEvent', 'App\Event\CustomListener');
```

I haven't had too much time to look through the Event Manager yet, but I wanted to start a discussion before I tried any code changes to ensure that any work would be on the right track.

Probably a 3.1 target release I'd think.

**Comment by jadb** [comment]
Isn't this already possible?

https://github.com/cakephp/cakephp/blob/3.0/src/Event/EventManager.php#L253

Maybe I misunderstood something?

**Comment by davidyell** [comment]
Yep, I had a chat with @lorenzo in irc about this and he said that it will only work with the same instance.

This is how I understand it.
```php
// TableClass.php

    // Attach listener
    $listener = new App\Event\CustomListener();
    $this->eventManager()->on($listener);
    
    // Remove listener
   $this->eventManager()->off($listener);
    // or
   $this->eventManager()->off('Some.Event, $listener);
```

Whereas this approach will not work, because the class is not the same instance.
```php
// MyBehaviour.php
    $listener = new App\Event\BehaviourListener();
    $this->eventManager()->on($listener);

// TableClass.php
    $listener = new App\Event\BehaviourListener();
    $this->eventManager()->off($listener);
```

**Comment by davidyell** [comment]
I've tried a small test and the behaviour matches my expectation as described by Jose.

**Comment by lorenzo** [comment]
I hope it is obvious why that cannot work: Having instances of the same class listening to events is quite possible, so you need the right instance in order to remove its listener.

**Comment by davidyell** [comment]
Sorry, but it's not obvious to me at all. I can't think of any use cases where I would want multiple instances of the same class listening to an event.

I think there must be a more elegant solution to this problem than removing all the listeners and re-attaching them all one by one.

**Comment by lorenzo** [comment]
I would ask, what is the need to remove a single listener key from an object?

**Comment by davidyell** [comment]
Using the events system provides a powerful way for plugin authors to give developers the ability to customise the functionality of a plugin without the need for lots of up front configuration. It also means that the plugin developer can select a sensible, even slightly opinionated, set of defaults for the plugin. If the chosen default configuration suits the developer then the plugin can run as normal, invisible to the developer, with the added bonus of very little configuration. Which is positive for plugin adoption due to the low barrier to entry created by a lack of confusing configuration options.

A developer looking to customise plugin functionality then has the option to hook the various events, as governed by the plugin author, without a need to change the internal api or extend classes, to customise the functionality. The plugin author can govern what is passed via the events when they are emitted. It's also leveraging the core because the events system is already there and creating an event listener is a simple process [documented well in the book](http://book.cakephp.org/3.0/en/core-libraries/events.html).

So in an instance where you had a number of listeners listening to events on your project already and you introduced a new plugin which allowed some customisation through the events system, I would certainly be put off by the idea of having to remove all my events, and rebind them all just to allow customisation of a plugin. In fact, I'd say it would be enough for me to find an alternative plugin. 

In summary I think that the events system provides huge opportunity for plugin authors, but the way in which it integrates with application code could be improved to add finer grained control to the Event Manager.

I'd still like to know if there is a use-case for having the same listener class listening more than once.

**Comment by markstory** [comment]
How can you remove a listener without the instance? Inferring which object to remove based on class name gets pretty hard with Closure.

**Comment by davidyell** [comment]
Wouldn't that just be a disadvantage for people using closures? What advantage would using a closure bring over the creation of an event listener class?

**Comment by markstory** [comment]
Closures are simpler to use and write. I often use closures for smaller tasks.
